mips: IRQ-protect stack ASLR randomization in arch_align_stack

namiltd opened the pull request:
On MT7620 (MIPS 24KEc), sporadic Oopses were observed during execve while randomizing the new process stack in arch_align_stack(), which calls get_random_u32_below(PAGE_SIZE).

On MIPS32, even the "<=16-bit" multiply path used by get_random_u32_below() emits a 32-bit multiply and reads HI/LO via mfhi/mflo. If a hard IRQ fires between multu and mfhi/mflo and its handler also executes mul/div, HI/LO can be clobbered, leading to corrupted intermediates and bogus addresses.

Make the random stack offset computation IRQ-safe on MT7620 by wrapping the get_random_u32_below(PAGE_SIZE) call with local_irq_save/restore(), ensuring the multu/mfhi/mflo sequence is not interrupted by hard IRQs. This preserves stack ASLR while eliminating the crash on this platform.

Fixes: https://github.com/openwrt/openwrt/issues/20298

namiltd commented:
Patch should also be used for the openWrt-24.10 branch.

neheb commented:
@neilbrown what do you think?

namiltd commented:
I've been searching for a solution to my MT7620 freezing issue for a long time. When I finally found a solution, not a workaround, I immediately shared it. I have no objections to extending this solution or even sending it upstream. I'm glad it finally works, so I can focus on the next issue: https://github.com/openwrt/openwrt/issues/20206

zx2c4 commented:
Nice detective work. I can't help but suspect this patch should be more general, though. Won't this be an issue for all multiplications? Maybe the IRQ handler should take care to not clobber these?

mrkiko commented:
Thanks a lot! Good work. And good to see people still taking care of older platforms. I apreciate it a lot.

zx2c4 commented:
Wondering if this is actually a compiler/cflags issue. The R4000 requires instructions between reading hi/lo and doing further multiplications, which hopefully the compiler would insert, unless this is being compiled for the wrong target?

namiltd commented:
The problem also occurs for other MIPS processors, so it makes sense to change the condition to:
`#if defined(CONFIG_MIPS) && !defined(CONFIG_CPU_MIPSR6)`
What do you think?

namiltd committed:
MIPS: IRQ-protect stack ASLR randomization in arch_align_stack

On MT7620 (MIPS 24KEc), sporadic Oopses were observed during execve while
randomizing the new process stack in arch_align_stack(), which calls
get_random_u32_below(PAGE_SIZE).

On MIPS32, even the "<=16-bit" multiply path used by get_random_u32_below()
emits a 32-bit multiply and reads HI/LO via mfhi/mflo. If a hard IRQ fires
between multu and mfhi/mflo and its handler also executes mul/div, HI/LO can
be clobbered, leading to corrupted intermediates and bogus addresses.

Make the random stack offset computation IRQ-safe on MT7620 by wrapping the
get_random_u32_below(PAGE_SIZE) call with local_irq_save/restore(), ensuring
the multu/mfhi/mflo sequence is not interrupted by hard IRQs. This preserves
stack ASLR while eliminating the crash on this platform.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>

namiltd commented:
Is there any chance of accepting this PR so that routers with mt7620 will finally stop hanging (also in branch 24.10)?

namiltd commented:
nstead of this patch you can use another one:
```
--- a/include/linux/random.h	2025-10-02 22:33:17.806249000 +0200
+++ b0/include/linux/random.h	2025-10-18 08:20:00.268892600 +0200
@@ -6,6 +6,7 @@
 #include <linux/bug.h>
 #include <linux/kernel.h>
 #include <linux/list.h>
+#include <linux/irqflags.h>
 
 #include <uapi/linux/random.h>
 
@@ -59,6 +60,7 @@
  */
 static inline u32 get_random_u32_below(u32 ceil)
 {
+	unsigned long flags;
 	if (!__builtin_constant_p(ceil))
 		return __get_random_u32_below(ceil);
 
@@ -73,19 +75,26 @@
 	BUILD_BUG_ON_MSG(!ceil, "get_random_u32_below() must take ceil > 0");
 	if (ceil <= 1)
 		return 0;
+	local_irq_save(flags);
 	for (;;) {
 		if (ceil <= 1U << 8) {
 			u32 mult = ceil * get_random_u8();
-			if (likely(is_power_of_2(ceil) || (u8)mult >= (1U << 8) % ceil))
+			if (likely(is_power_of_2(ceil) || (u8)mult >= (1U << 8) % ceil)) {
+				local_irq_restore(flags);
 				return mult >> 8;
+			}
 		} else if (ceil <= 1U << 16) {
 			u32 mult = ceil * get_random_u16();
-			if (likely(is_power_of_2(ceil) || (u16)mult >= (1U << 16) % ceil))
+			if (likely(is_power_of_2(ceil) || (u16)mult >= (1U << 16) % ceil)) {
+				local_irq_restore(flags);
 				return mult >> 16;
+			}
 		} else {
 			u64 mult = (u64)ceil * get_random_u32();
-			if (likely(is_power_of_2(ceil) || (u32)mult >= -ceil % ceil))
+			if (likely(is_power_of_2(ceil) || (u32)mult >= -ceil % ceil)) {
+				local_irq_restore(flags);
 				return mult >> 32;
+			}
 		}
 	}
 }
```
Any changes inside get_random_## (get_random_u32) have no effect.

zx2c4 commented:
> Is there any chance of accepting this PR so that routers with mt7620 will finally stop hanging (also in branch 24.10)?

IMO, no, this is not the right fix.
